From ca45009cd43cec3c30874490fb753ef08d3a0c45 Mon Sep 17 00:00:00 2001
From: Gong S <42509223+pthfdr-42@users.noreply.github.com>
Date: Wed, 6 May 2020 12:20:05 +0800
Subject: [PATCH] Change mount options for some FSes.

1. `relatime` is the default option in current distros. `noatime` is not really necessary.
2. Permissionless FSes should not leak execute permissions.
3. Add edge cases about NTFS and BTRFS.
---
 etc/udevil.conf | 43 ++++++++++++++++++++++++-------------------
 1 file changed, 24 insertions(+), 19 deletions(-)

diff --git a/etc/udevil.conf b/etc/udevil.conf
index 75a4c0d..08a8b60 100644
--- a/etc/udevil.conf
+++ b/etc/udevil.conf
@@ -101,7 +101,7 @@ allowed_groups = *
 # be permitted to unmount its associated loop device even though internal.
 # INCLUDING /MNT HERE IS NOT RECOMMENDED.  ALL ALLOWED MEDIA DIRECTORIES
 # SHOULD BE OWNED AND WRITABLE ONLY BY ROOT.
-allowed_media_dirs = /media, /media/$USER, /run/media/$USER
+allowed_media_dirs = /run/media/$USER, /run/mount/$USER, /media/$USER, /run/media, /run/mount, /media, /mnt
 
 
 # allowed_devices is the first criteria for what block devices users may mount
@@ -216,23 +216,24 @@ forbidden_files =
 # FOR GOOD SECURITY, default_options SHOULD ALWAYS INCLUDE: nosuid,noexec,nodev
 # WARNING:  OPTIONS PRESENT OR MISSING CAN CAUSE SERIOUS SECURITY PROBLEMS.
 default_options           = nosuid, noexec, nodev, noatime
-default_options_file      = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID, ro
+default_options_file      = nosuid, noexec, nodev, uid=$UID, gid=$GID
 # mount iso9660 with 'ro' to prevent mount read-only warning
-default_options_iso9660   = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID, ro, utf8
-default_options_udf       = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID
-default_options_vfat      = nosuid, noexec, nodev, noatime, fmask=0133, dmask=0022, uid=$UID, gid=$GID, utf8
-default_options_exfat     = nosuid, noexec, nodev, noatime, umask=0077, uid=$UID, gid=$GID, iocharset=utf8, namecase=0, nonempty
-default_options_msdos     = nosuid, noexec, nodev, noatime, fmask=0133, dmask=0022, uid=$UID, gid=$GID
-default_options_umsdos    = nosuid, noexec, nodev, noatime, fmask=0133, dmask=0022, uid=$UID, gid=$GID
-default_options_ntfs      = nosuid, noexec, nodev, noatime, fmask=0133, uid=$UID, gid=$GID, utf8
+default_options_iso9660   = nosuid, noexec, nodev, uid=$UID, gid=$GID, ro, utf8
+default_options_udf       = nosuid, noexec, nodev, uid=$UID, gid=$GID
+default_options_vfat      = nosuid, noexec, nodev, uid=$UID, gid=$GID, fmask=7177, dmask=7077, utf8
+default_options_exfat     = nosuid, noexec, nodev, uid=$UID, gid=$GID, fmask=7177, dmask=7077, iocharset=utf8, namecase=0, nonempty
+default_options_msdos     = nosuid, noexec, nodev, uid=$UID, gid=$GID, fmask=7177, dmask=7077
+default_options_umsdos    = nosuid, noexec, nodev, uid=$UID, gid=$GID, fmask=7177, dmask=7077
+default_options_ntfs      = nosuid, noexec, nodev, uid=$UID, gid=$GID, fmask=7177, dmask=7077, utf8, show_sys_files, windows_names, big_writes
+default_options_btrfs     = nosuid, noexec, nodev, autodefrag
 default_options_cifs      = nosuid, noexec, nodev, uid=$UID, gid=$GID
 default_options_smbfs     = nosuid, noexec, nodev, uid=$UID, gid=$GID
-default_options_sshfs     = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID, nonempty, allow_other
-default_options_curlftpfs = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID, nonempty, allow_other
-default_options_ftpfs     = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID
+default_options_sshfs     = nosuid, noexec, nodev, uid=$UID, gid=$GID, nonempty, allow_other
+default_options_curlftpfs = nosuid, noexec, nodev, uid=$UID, gid=$GID, nonempty, allow_other
+default_options_ftpfs     = nosuid, noexec, nodev, uid=$UID, gid=$GID
 default_options_davfs     = nosuid, noexec, nodev, uid=$UID, gid=$GID
-default_options_tmpfs     = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID
-default_options_ramfs     = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID
+default_options_tmpfs     = nosuid, noexec, nodev, uid=$UID, gid=$GID
+default_options_ramfs     = nosuid, noexec, nodev, uid=$UID, gid=$GID
 
 
 # allowed_options determines all options that a user may specify when mounting.
@@ -245,15 +246,19 @@ default_options_ramfs     = nosuid, noexec, nodev, noatime, uid=$UID, gid=$GID
 # and GID.
 # If you want to forbid remounts, remove 'remount' from here.
 # WARNING:  OPTIONS HERE CAN CAUSE SERIOUS SECURITY PROBLEMS - CHOOSE CAREFULLY
-allowed_options           = nosuid, noexec, nodev, noatime, fmask=0133, dmask=0022, uid=$UID, gid=$GID, ro, rw, sync, flush, iocharset=*, utf8, remount
+allowed_options           = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, uid=$UID, gid=$GID, ro, rw, sync, flush, iocharset=*, utf8, nobarrier, remount
 allowed_options_nfs       = nosuid, noexec, nodev, noatime, ro, rw, sync, remount, port=*, rsize=*, wsize=*, hard, proto=*, timeo=*, retrans=*
-allowed_options_cifs      = nosuid, noexec, nodev, ro, rw, remount, port=*, user=*, username=*, pass=*, password=*, guest, domain=*, uid=$UID, gid=$GID, credentials=*
-allowed_options_smbfs     = nosuid, noexec, nodev, ro, rw, remount, port=*, user=*, username=*, pass=*, password=*, guest, domain=*, uid=$UID, gid=$GID, credentials=*
+allowed_options_cifs      = nosuid, noexec, nodev, noatime, ro, rw, remount, port=*, user=*, username=*, pass=*, password=*, guest, domain=*, uid=$UID, gid=$GID, credentials=*
+allowed_options_smbfs     = nosuid, noexec, nodev, noatime, ro, rw, remount, port=*, user=*, username=*, pass=*, password=*, guest, domain=*, uid=$UID, gid=$GID, credentials=*
 allowed_options_sshfs     = nosuid, noexec, nodev, noatime, ro, rw, uid=$UID, gid=$GID, nonempty, allow_other, idmap=user, BatchMode=yes, port=*
 allowed_options_curlftpfs = nosuid, noexec, nodev, noatime, ro, rw, uid=$UID, gid=$GID, nonempty, allow_other, user=*
 allowed_options_ftpfs     = nosuid, noexec, nodev, noatime, ro, rw, port=*, user=*, pass=*, root=*, uid=$UID, gid=$GID
-allowed_options_exfat     = nosuid, noexec, nodev, noatime, fmask=0133, dmask=0022, uid=$UID, gid=$GID, umask=0077, namecase=*, ro, rw, sync, flush, iocharset=*, remount, nonempty
-
+allowed_options_exfat     = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, uid=$UID, gid=$GID, umask=*, namecase=*, ro, rw, sync, flush, iocharset=*, remount, nonempty
+allowed_options_ntfs      = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, uid=$UID, gid=$GID, show_sys_files, windows_names, compression, nocompression, big_writes, remount
+allowed_options_vfat      = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, uid=$UID, gid=$GID, ro, rw, sync, flush, iocharset=*, utf8, remount, check=*, codepage=*
+allowed_options_msdos     = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, uid=$UID, gid=$GID, ro, rw, sync, flush, iocharset=*, utf8, remount, check=*, codepage=*
+allowed_options_umsdos    = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, uid=$UID, gid=$GID, ro, rw, sync, flush, iocharset=*, utf8, remount, check=*, codepage=*
+allowed_options_btrfs     = nosuid, noexec, nodev, noatime, fmask=*, dmask=*, ro, rw, sync, flush, autodefrag, nobarrier, compress=*, compress-force=*, nodatacow, subvol=*, subvolid=*, remount
 
 # mount_point_mode, if present and set to a non-empty value, will cause udevil
 # to set the mode (permissions) on the moint point after mounting  If not
